File: app/src/main/java/com/aplicacion/mypet/activities/publicar/ActivityCrearPublicacion.java
package com.aplicacion.mypet.activities.publicar;

import android.Manifest;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.FileProvider;

import com.aplicacion.mypet.R;
import com.aplicacion.mypet.fragments.BottomSheetFragmentPersonalizado;
import com.aplicacion.mypet.models.Publicacion;
import com.aplicacion.mypet.providers.AuthProvider;
import com.aplicacion.mypet.providers.ImageProvider;
import com.aplicacion.mypet.providers.PublicacionProvider;
import com.aplicacion.mypet.utils.FileUtil;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.storage.UploadTask;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;

import dmax.dialog.SpotsDialog;

public class ActivityCrearPublicacion extends AppCompatActivity implements BottomSheetFragmentPersonalizado.BottomSheetListener {
    private final int GALLERY_REQUEST_CODE=1;
    private final int CAMARA_REQUEST_CODE=2;
    private final int REQUEST_PERMISSION_CAMERA=100;
    private final int REQUEST_PERMISSION_GALLERY=101;

    private AuthProvider authProvider;
    private int contador=0;


    private PublicacionProvider publicacionProvider;
    private File[] imagenes;
    private ArrayList<File> imagenesAlmacenadas;
    private ImageProvider imageProvider;
    private ImageView imagenSeleccionada;
    private File fotoCamaraArchivo;

    private Button boton_tipo;
    private TextInputEditText inputTextNombre;
    private TextInputEditText inputTextEdad;
    private TextInputEditText inputTextRaza;
    private TextInputEditText inputTextDescripcion;
    private int posicionArraytipo;

    private String nombre;
    private String edad;
    private String tipo;
    private String raza;
    private String descripcion;
    private String sexo;
    private ArrayList<String> urlImagenes;

    private AlertDialog dialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_crear_publicacion);

        imageProvider = new ImageProvider();
        sexo="";
        imagenes = new File[5];

        boton_tipo = findViewById(R.id.boton_cambiar_tipo);
        inputTextNombre = findViewById(R.id.campo_nombre_animal);
        inputTextEdad = findViewById(R.id.campo_edad);
        inputTextRaza = findViewById(R.id.campo_raza);
        inputTextDescripcion = findViewById(R.id.campo_descripcion);


        urlImagenes = new ArrayList<>();
        publicacionProvider = new PublicacionProvider();

        authProvider = new AuthProvider();

        dialog = new SpotsDialog.Builder()
                .setContext(this)
                .setMessage(R.string.subiendo_publicacion)
                .setCancelable(false).build();
    }

    public void cancelarPublicacion(View v) {
        finish();
    }

    public void cambiarTipoAnimal(View view) {
        AlertDialog dialogo;
        String[] array = getResources().getStringArray(R.array.lista_animales);
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(getString(R.string.tipo_animal))
                .setSingleChoiceItems(array,0, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        boton_tipo.setText(array[which]);
                        posicionArraytipo = which;
                        //boton_tipo.setTextColor(getResources().getColor(R.color.black));
                        dialog.cancel();
                    }
                });

        dialogo = builder.create();
        dialogo.show();
    }

    public void elegirImagen(View view) {
        imagenSeleccionada = (ImageView) view;
        BottomSheetFragmentPersonalizado bottomSheet = new BottomSheetFragmentPersonalizado();
        bottomSheet.show(getSupportFragmentManager(),"TAG");
    }




    public void subirPublicacion(View view) {
        nombre = inputTextNombre.getText().toString();
        edad = inputTextEdad.getText().toString();
        tipo = boton_tipo.getText().toString();
        raza = inputTextRaza.getText().toString();
        descripcion = inputTextDescripcion.getText().toString();


        if (!nombre.isEmpty() && !edad.isEmpty() && !tipo.equals(getString(R.string.tipo_animal))
            && !raza.isEmpty() && !sexo.isEmpty() && !descripcion.isEmpty() && imagenesAlmacenadas.size()>0) {
            dialog.show();
            saveImage();
        } else {
            Toast.makeText(ActivityCrearPublicacion.this, getString(R.string.campos_vacios), Toast.LENGTH_LONG).show();
        }

    }



    public void seleccionarSexo(View view) {
        ImageView imagenMasculino = findViewById(R.id.masculino);
        ImageView imagenFemenino = findViewById(R.id.femenino);
        LinearLayout linearSexo = (LinearLayout) view;
        if (linearSexo.getId() == R.id.sexo_masculino) {
            imagenFemenino.setBackgroundColor(getResources().getColor(R.color.white,getTheme()));
            imagenMasculino.setBackgroundColor(getResources().getColor(R.color.principal_app,getTheme()));
        } else if (linearSexo.getId() == R.id.sexo_femenino) {
            imagenMasculino.setBackgroundColor(getResources().getColor(R.color.white,getTheme()));
            imagenFemenino.setBackgroundColor(getResources().getColor(R.color.principal_app,getTheme()));
        }
        sexo = (String) linearSexo.getContentDescription();
    }

    private void saveImage() {
        if (contador < imagenesAlmacenadas.size()) {
            contador++;
            imageProvider.save(this,imagenesAlmacenadas.get(contador-1),contador-1).addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
                @Override
                public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {
                    imageProvider.getStorage().addOnSuccessListener(new OnSuccessListener<Uri>() {
                        @Override
                        public void onSuccess(Uri uri) {
                            urlImagenes.add(uri.toString());
                            if (contador == imagenesAlmacenadas.size()) {

                                Publicacion publicacion = new Publicacion();
                                publicacion.setNombre(nombre.trim());
                                publicacion.setEdad(edad.trim());
                                publicacion.setTipo(String.valueOf(posicionArraytipo));
                                publicacion.setRaza(raza.trim());
                                publicacion.setDescripcion(descripcion.trim());
                                publicacion.setSexo(sexo.trim());
                                publicacion.setFechaPublicacion(new Date().getTime());
                                publicacion.setIdUser(authProvider.getUid());
                                publicacion.setImagenes(urlImagenes);

                                publicacionProvider.save(publicacion).addOnCompleteListener(new OnCompleteListener<Void>() {
                                    @Override
                                    public void onComplete(@NonNull Task<Void> taskSave) {
                                        if (taskSave.isSuccessful()) {
                                            dialog.dismiss();
                                            Toast.makeText(ActivityCrearPublicacion.this, getString(R.string.animal_publicado), Toast.LENGTH_LONG).show();
                                            finish();
                                        }else {
                                            Toast.makeText(ActivityCrearPublicacion.this, getString(R.string.error_animal_publicado), Toast.LENGTH_LONG).show();
                                        }
                                    }
                                });
                            }
                            saveImage();
                        }
                    });
                }
            });
        }
    }

    private void openGallery() {
        Intent galleryIntent = new Intent(Intent.ACTION_GET_CONTENT);
        galleryIntent.setType("image/*");
        startActivityForResult(galleryIntent,GALLERY_REQUEST_CODE);
    }

    private void openCamera() {
        Intent cameraIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        if(cameraIntent.resolveActivity(getPackageManager())!=null){

            fotoCamaraArchivo = null;
            try {
                fotoCamaraArchivo = FileUtil.fileCamera(this);
            } catch (IOException e) {
                e.printStackTrace();
            }

            if (fotoCamaraArchivo != null) {
                Uri fotoUri = FileProvider.getUriForFile(
                        this,
                        "com.aplicacion.mypet",
                        fotoCamaraArchivo);
                cameraIntent.putExtra(MediaStore.EXTRA_OUTPUT, fotoUri);
                startActivityForResult(cameraIntent,CAMARA_REQUEST_CODE);
            }
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode==GALLERY_REQUEST_CODE && resultCode == RESULT_OK) {
            try {
                File imagen = FileUtil.from(this,data.getData());
                asignarImagenesArray(imagen);
                imagenSeleccionada.setImageBitmap(BitmapFactory.decodeFile(imagen.getAbsolutePath()));
            }catch (Exception e){
                Log.d("ERROR: ", e.getMessage());
                Toast.makeText(this, "Error: " + e.getMessage(), Toast.LENGTH_SHORT).show();
            }
        } else if (requestCode == CAMARA_REQUEST_CODE && resultCode == RESULT_OK){
            asignarImagenesArray(fotoCamaraArchivo);

            imagenSeleccionada.setImageBitmap(BitmapFactory.decodeFile(fotoCamaraArchivo.getAbsolutePath()));
        }
    }

    private void asignarImagenesArray(File file) {
        if (imagenSeleccionada.getId() == R.id.imagen_publicar_1){
            imagenes[0] = file;
        } else if (imagenSeleccionada.getId() == R.id.imagen_publicar_2){
            imagenes[1] = file;
        } else if (imagenSeleccionada.getId() == R.id.imagen_publicar_3){
            imagenes[2] = file;
        } else if (imagenSeleccionada.getId() == R.id.imagen_publicar_4){
            imagenes[3] = file;
        } else if (imagenSeleccionada.getId() == R.id.imagen_publicar_5){
            imagenes[4] = file;
        }
        actualizarArrayListImagenes();
    }

    private void eliminarImagenesArray() {
        if (imagenSeleccionada.getId() == R.id.imagen_publicar_1){
            imagenes[0] = null;
        } else if (imagenSeleccionada.getId() == R.id.imagen_publicar_2){
            imagenes[1] = null;
        } else if (imagenSeleccionada.getId() == R.id.imagen_publicar_3){
            imagenes[2] = null;
        } else if (imagenSeleccionada.getId() == R.id.imagen_publicar_4){
            imagenes[3] = null;
        } else if (imagenSeleccionada.getId() == R.id.imagen_publicar_5){
            imagenes[4] = null;
        }
        actualizarArrayListImagenes();
    }

    private void actualizarArrayListImagenes() {
        imagenesAlmacenadas = new ArrayList<>();
        for (File imagen : imagenes) {
            if (imagen!= null)
                imagenesAlmacenadas.add(imagen);
        }
    }

    @Override
    public void onButtonClicked(int numero) {
        if (numero == 0){
            if(ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED){
                openCamera();
            } else {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA},REQUEST_PERMISSION_CAMERA);
            }
        } else if (numero == 1) {
            if(ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED){
                openGallery();
            } else {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},REQUEST_PERMISSION_GALLERY);
            }
        } else if (numero == 2) {
            imagenSeleccionada.setImageResource(R.drawable.ic_camara);
            eliminarImagenesArray();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if(requestCode == REQUEST_PERMISSION_CAMERA) {
            if (permissions.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED){
                openCamera();
            } else {
                Toast.makeText(ActivityCrearPublicacion.this, getString(R.string.permisos), Toast.LENGTH_LONG).show();
            }
        } else if(requestCode == REQUEST_PERMISSION_GALLERY) {
            if (permissions.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED){
                openGallery();
            } else {
                Toast.makeText(ActivityCrearPublicacion.this, getString(R.string.permisos), Toast.LENGTH_LONG).show();
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }
}

File: app/src/main/java/com/aplicacion/mypet/activities/publicar/ActivityCrearPublicacion.kt
package com.aplicacion.mypet.activities.publicar

import android.Manifest
import android.app.AlertDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.provider.MediaStore
import android.util.Base64
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.FileProvider
import com.aplicacion.mypet.R
import com.aplicacion.mypet.fragments.BottomSheetFragmentPersonalizado
import com.aplicacion.mypet.models.Publicacion
import com.aplicacion.mypet.providers.AuthProvider
import com.aplicacion.mypet.providers.ImageProvider
import com.aplicacion.mypet.providers.PublicacionProvider
import com.aplicacion.mypet.utils.FileUtil
import com.aplicacion.mypet.utils.ViewedMessageHelper
import com.google.android.material.textfield.TextInputEditText
import dmax.dialog.SpotsDialog
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.*
import kotlin.collections.ArrayList

class ActivityCrearPublicacion : AppCompatActivity(),BottomSheetFragmentPersonalizado.BottomSheetListener {
    private val REQUEST_PERMISSION_CAMERA = 100
    private val REQUEST_PERMISSION_GALLERY = 101

    private var authProvider: AuthProvider? = null
    private var contador = 0



    private lateinit var publicacionProvider: PublicacionProvider
    private lateinit var imagenes: Array<File?>
    private lateinit var imagenesAlmacenadas: ArrayList<File?>
    private lateinit var imageProvider: ImageProvider
    private lateinit var imagenSeleccionada: ImageView
    private var fotoCamaraArchivo: File? = null

    private lateinit var cuadrosDeImagenes: ArrayList<ImageView>

    private lateinit var botonTipo: Button
    private lateinit var inputTextNombre: TextInputEditText
    private lateinit var inputTextEdad: TextInputEditText
    private lateinit var inputTextRaza: TextInputEditText
    private lateinit var inputTextDescripcion: TextInputEditText
    private var posicionArraytipo = 0

    private lateinit var nombre: String
    private lateinit var edad: String
    private lateinit var tipo: String
    private lateinit var raza: String
    private lateinit var descripcion: String
    private lateinit var sexo: String
    private lateinit var urlImagenes: ArrayList<String>

    private var idPublicacion : String? = null

    private lateinit var dialog: AlertDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_crear_publicacion)

        imageProvider = ImageProvider()
        authProvider = AuthProvider()
        publicacionProvider = PublicacionProvider()

        idPublicacion = intent.extras?.getString("idPublicacion").toString()

        cuadrosDeImagenes = ArrayList()
        cuadrosDeImagenes.add(findViewById(R.id.imagen_publicar_1))
        cuadrosDeImagenes.add(findViewById(R.id.imagen_publicar_2))
        cuadrosDeImagenes.add(findViewById(R.id.imagen_publicar_3))
        cuadrosDeImagenes.add(findViewById(R.id.imagen_publicar_4))
        cuadrosDeImagenes.add(findViewById(R.id.imagen_publicar_5))

        sexo =""
        imagenes = arrayOfNulls(5)

        botonTipo = findViewById(R.id.boton_cambiar_tipo)
        inputTextNombre = findViewById(R.id.campo_nombre_animal)
        inputTextEdad = findViewById(R.id.campo_edad)
        inputTextRaza = findViewById(R.id.campo_raza)
        inputTextDescripcion = findViewById(R.id.campo_descripcion)



        urlImagenes = ArrayList()
        imagenesAlmacenadas = ArrayList()

        dialog = SpotsDialog.Builder()
                .setContext(this)
                .setMessage(R.string.subiendo_publicacion)
                .setCancelable(false).build()


        if (idPublicacion != null) {
            getPublicacion()
        }
    }

    fun cancelarPublicacion(v: View?) {
        finish()
    }

    fun getPublicacion() {
        publicacionProvider.getPostById(idPublicacion).addOnSuccessListener { documentSnapshot ->
            if (documentSnapshot.contains("imagenes")) {
                urlImagenes = documentSnapshot["imagenes"] as ArrayList<String>
                for (i in urlImagenes.indices) {

                    imageProvider.getImageFile(urlImagenes[i]).addOnSuccessListener { task ->

                        var bitmap = BitmapFactory.decodeByteArray(task, 0, task.size)
                        cuadrosDeImagenes[i].setImageBitmap(bitmap)
                        imagenSeleccionada = cuadrosDeImagenes[i]

                        val f: File = File.createTempFile("imagen$i", ".jpg")
                        f.writeBytes(task)
                        asignarImagenesArray(f)
                    }
                }
            }

            if (documentSnapshot.contains("nombre")) {
                inputTextNombre.setText(documentSnapshot.getString("nombre"))
            }

            if (documentSnapshot.contains("edad")) {
                inputTextEdad.setText(documentSnapshot.getString("edad"))
            }

            if (documentSnapshot.contains("raza")) {
                inputTextRaza.setText(documentSnapshot.getString("raza"))
            }

            if (documentSnapshot.contains("tipo")) {
                val tipo = documentSnapshot.getString("tipo")!!.toInt()
                botonTipo.text = resources.getStringArray(R.array.lista_animales)[tipo]
            }

            if (documentSnapshot.contains("descripcion")) {
                inputTextDescripcion.setText(documentSnapshot.getString("descripcion"))
            }

            if (documentSnapshot.contains("sexo")) {
                val imagenMasculino: ImageView = findViewById(R.id.masculino)
                val imagenFemenino: ImageView = findViewById(R.id.femenino)

                sexo = documentSnapshot.getString("sexo").toString()
                if (sexo.equals(getString(R.string.desc_femenino), ignoreCase = true)) {
                    imagenFemenino.setBackgroundColor(resources.getColor(R.color.white, theme))
                    imagenMasculino.setBackgroundColor(resources.getColor(R.color.principal_app, theme))
                } else if (sexo.equals(getString(R.string.desc_masculino), ignoreCase = true)) {
                    imagenMasculino.setBackgroundColor(resources.getColor(R.color.white, theme))
                    imagenFemenino.setBackgroundColor(resources.getColor(R.color.principal_app, theme))
                }


            }
        }
    }

    fun cambiarTipoAnimal(view: View?) {
        val dialogo: AlertDialog
        val array = resources.getStringArray(R.array.lista_animales)
        val builder = AlertDialog.Builder(this)
        builder.setTitle(getString(R.string.tipo_animal))
                .setSingleChoiceItems(array, 0) { dialog, which ->
                    botonTipo.text = array[which]
                    posicionArraytipo = which
                    dialog.cancel()
                }

        dialogo = builder.create()
        dialogo.show()
    }

    fun elegirImagen(view: View) {
        imagenSeleccionada = view as ImageView
        val bottomSheet = BottomSheetFragmentPersonalizado()
        bottomSheet.show(supportFragmentManager, "TAG")
    }

    fun seleccionarSexo(view: View?) {
        val imagenMasculino: ImageView = findViewById(R.id.masculino)
        val imagenFemenino: ImageView = findViewById(R.id.femenino)

        val linearSexo = view as LinearLayout

        if (linearSexo.id == R.id.sexo_masculino) {
            imagenFemenino.setBackgroundColor(resources.getColor(R.color.white, theme))
            imagenMasculino.setBackgroundColor(resources.getColor(R.color.principal_app, theme))
        } else if (linearSexo.id == R.id.sexo_femenino) {
            imagenMasculino.setBackgroundColor(resources.getColor(R.color.white, theme))
            imagenFemenino.setBackgroundColor(resources.getColor(R.color.principal_app, theme))
        }

        sexo = linearSexo.contentDescription as String
    }

    fun subirPublicacion(view: View?) {
        nombre = inputTextNombre.text.toString()
        edad = inputTextEdad.text.toString()
        tipo = botonTipo.text.toString()
        raza = inputTextRaza.text.toString()
        descripcion = inputTextDescripcion.text.toString()

        if(nombre.isNotEmpty() && edad.isNotEmpty() && !tipo.equals(getString(R.string.tipo_animal))
                && raza.isNotEmpty() && sexo.isNotEmpty() && descripcion.isNotEmpty() && imagenesAlmacenadas.size > 0) {
            dialog.show()
            if (idPublicacion == "null") {
                saveImages()
            } else {
                actualizarPublicacion()
            }
        } else {
            Toast.makeText(this@ActivityCrearPublicacion, getString(R.string.campos_vacios), Toast.LENGTH_LONG).show()
        }
    }

    private fun saveImages() {
        if (contador < imagenesAlmacenadas.size) {
            contador++
            imageProvider.save(this, imagenesAlmacenadas.get(contador - 1), contador - 1).addOnSuccessListener { taskSnapshot ->
                imageProvider.storage.addOnSuccessListener { uri ->
                    urlImagenes.add(uri.toString())

                    if (contador == imagenesAlmacenadas.size) {
                        val publicacion = Publicacion()

                        publicacion.nombre = nombre.trim()
                        publicacion.edad = edad.trim()
                        publicacion.tipo = posicionArraytipo.toString()
                        publicacion.raza = raza.trim()
                        publicacion.descripcion = descripcion.trim()
                        publicacion.sexo = sexo.trim()
                        publicacion.fechaPublicacion = Date().time
                        publicacion.idUser = authProvider!!.uid
                        publicacion.imagenes = urlImagenes

                        publicacionProvider.save(publicacion).addOnCompleteListener { taskSave ->
                            if (taskSave.isSuccessful) {
                                dialog.dismiss()
                                Toast.makeText(this@ActivityCrearPublicacion, getString(R.string.animal_publicado), Toast.LENGTH_LONG).show()
                                finish()
                            } else {
                                Toast.makeText(this@ActivityCrearPublicacion, getString(R.string.error_animal_publicado), Toast.LENGTH_LONG).show()
                            }
                        }
                    }
                    saveImages()
                }
            }
        }
    }

    private fun actualizarPublicacion(){
        for (i in urlImagenes!!.indices) {
            imageProvider.deleteByUrl(urlImagenes!![i])
        }
        urlImagenes.clear()
        saveImageUpdate()
    }

    private fun saveImageUpdate() {
        if (contador < imagenesAlmacenadas.size) {
            contador++
            imageProvider.save(this, imagenesAlmacenadas[contador - 1], contador - 1).addOnSuccessListener { taskSnapshot ->
                imageProvider.storage.addOnSuccessListener { uri ->
                    urlImagenes.add(uri.toString())

                    if (contador == imagenesAlmacenadas.size) {
                        val publicacion = Publicacion()

                        publicacion.nombre = nombre.trim()
                        publicacion.edad = edad.trim()
                        publicacion.tipo = posicionArraytipo.toString()
                        publicacion.raza = raza.trim()
                        publicacion.descripcion = descripcion.trim()
                        publicacion.sexo = sexo.trim()
                        publicacion.id = idPublicacion
                        publicacion.fechaPublicacion = Date().time
                        publicacion.imagenes = urlImagenes

                        publicacionProvider.update(publicacion).addOnCompleteListener { taskSave ->
                            if (taskSave.isSuccessful) {
                                dialog.dismiss()
                                Toast.makeText(this@ActivityCrearPublicacion, getString(R.string.publicacion_actualizada), Toast.LENGTH_LONG).show()
                                finish()
                            } else {
                                Toast.makeText(this@ActivityCrearPublicacion, getString(R.string.error_animal_publicado), Toast.LENGTH_LONG).show()
                            }
                        }
                    }
                    saveImageUpdate()
                }
            }
        }
    }

    private fun openGallery() {
        val galleryIntent = Intent(Intent.ACTION_GET_CONTENT)
        galleryIntent.type = "image/*"
        resultGaleria.launch(galleryIntent)
    }

    private fun openCamera() {
        val cameraIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (cameraIntent.resolveActivity(packageManager) != null) {
            try {
                fotoCamaraArchivo = FileUtil.fileCamera(this)
            } catch (e: IOException) {
                e.printStackTrace()
            }
            if (fotoCamaraArchivo != null) {
                val fotoUri = FileProvider.getUriForFile(
                        this,
                        "com.aplicacion.mypet",
                        fotoCamaraArchivo!!)
                cameraIntent.putExtra(MediaStore.EXTRA_OUTPUT, fotoUri)
                resultCamara.launch(cameraIntent)
            }
        }
    }

    var resultGaleria = registerForActivityResult(
            StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            try {
                val imagen = FileUtil.from(this@ActivityCrearPublicacion, result.data!!.data)
                asignarImagenesArray(imagen)
                imagenSeleccionada.setImageBitmap(BitmapFactory.decodeFile(imagen.absolutePath))
            } catch (e: Exception) {
                Log.d("ERROR: ", e.message!!)
                Toast.makeText(this@ActivityCrearPublicacion, "Error: " + e.message, Toast.LENGTH_SHORT).show()
            }
        }
    }

    var resultCamara = registerForActivityResult(
            StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            asignarImagenesArray(fotoCamaraArchivo!!)
            imagenSeleccionada.setImageBitmap(BitmapFactory.decodeFile(fotoCamaraArchivo!!.absolutePath))
        }
    }

    private fun asignarImagenesArray(file: File) {
        if (imagenSeleccionada.id == R.id.imagen_publicar_1) {
            imagenes[0] = file
        } else if (imagenSeleccionada.id == R.id.imagen_publicar_2) {
            imagenes[1] = file
        } else if (imagenSeleccionada.id == R.id.imagen_publicar_3) {
            imagenes[2] = file
        } else if (imagenSeleccionada.id == R.id.imagen_publicar_4) {
            imagenes[3] = file
        } else if (imagenSeleccionada.id == R.id.imagen_publicar_5) {
            imagenes[4] = file
        }
        actualizarArrayListImagenes()
    }

    private fun eliminarImagenesArray() {
        if (imagenSeleccionada.id == R.id.imagen_publicar_1) {
            imagenes[0] = null
        } else if (imagenSeleccionada.id == R.id.imagen_publicar_2) {
            imagenes[1] = null
        } else if (imagenSeleccionada.id == R.id.imagen_publicar_3) {
            imagenes[2] = null
        } else if (imagenSeleccionada.id == R.id.imagen_publicar_4) {
            imagenes[3] = null
        } else if (imagenSeleccionada.id == R.id.imagen_publicar_5) {
            imagenes[4] = null
        }
        actualizarArrayListImagenes()
    }

    private fun actualizarArrayListImagenes() {
        imagenesAlmacenadas = ArrayList()
        for (imagen in imagenes) {
            if (imagen != null) imagenesAlmacenadas.add(imagen)
        }
    }

    override fun onButtonClicked(numero: Int) {
        if (numero == 0) {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
                openCamera()
            } else {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA), REQUEST_PERMISSION_CAMERA)
            }
        } else if (numero == 1) {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                openGallery()
            } else {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), REQUEST_PERMISSION_GALLERY)
            }
        } else if (numero == 2) {
            imagenSeleccionada.setImageResource(R.mipmap.ic_subir_foto)
            eliminarImagenesArray()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String?>, grantResults: IntArray) {
        if (requestCode == REQUEST_PERMISSION_CAMERA) {
            if (permissions.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                openCamera()
            } else {
                Toast.makeText(this@ActivityCrearPublicacion, getString(R.string.permisos), Toast.LENGTH_LONG).show()
            }
        } else if (requestCode == REQUEST_PERMISSION_GALLERY) {
            if (permissions.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                openGallery()
            } else {
                Toast.makeText(this@ActivityCrearPublicacion, getString(R.string.permisos), Toast.LENGTH_LONG).show()
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onStart() {
        super.onStart()
        ViewedMessageHelper.updateOnline(true, this@ActivityCrearPublicacion)
    }


    override fun onPause() {
        super.onPause()
        ViewedMessageHelper.updateOnline(false, this)
    }
}